Workflows/Rails upgrades.md

```diff
@@ -0,0 +1,286 @@
+Rails is probably one of your biggest project dependencies. While it may look like a big blob of magic, Rails is in essence a composition of multiple opinionated ([omakase](https://dhh.dk/2012/rails-is-omakase.html)) frameworks. Each of the individual functionalities (object-relation mapping, view templating, email sending, ...) is handled by its own gem:
+
+`actioncable, actionmailbox, actionmailer, actionpack, actiontext, actionview, activejob, activemodel, activerecord, activestorage, activesupport`
+
+Upgrading Rails is thus much more [challenging](https://engineering.shopify.com/blogs/engineering/upgrading-shopify-to-rails-5-0) and [tends to hurt](https://blog.testdouble.com/posts/2019-09-03-3-keys-to-upgrading-rails) because of the big surface area. You're essentially upgrading 11 gems at one time, which are themselves dependencies to a lot of your project's dependencies.
+
+
+## Upgrade strategy
+Planning a Rails upgrade strategy is trivial on well maintained projects where we're only one minor/major Rails version behind, but there are projects where we're severly behind and need to set a precise plan of action. The steps below can guide you in both of those cases and should be formally noted as part of the upgrade task.
+
+### Assessing the current state of the project
+The prerequisite for creating a roadmap is knowing your current location. You should make note of these properties before commencing a Rails upgrade cycle:
+
+- Ruby version
+  - how many minor/major versions away from the latest version is it
+  - is it [EOL](https://endoflife.date/ruby)
+- Rails version
+  - how many minor/major versions away from the latest version is it
+  - is it [EOL](https://endoflife.date/rails)
+- test suite confidence
+  - number of tests
+  - code coverage
+  - types of tests
+  - do tests pass locally
+- deployment pipeline
+  - is it functional
+  - do we have access to it
+- dependency age
+  - how many dependencies does the project have
+  - when were they last updated
+  - how far away from the latest version are they
+- context
+  - do we have anyone from the client side available to answer basic questions
+  - do we have anyone with in-depth knowledge of the project available internally
+  - does the QA team have a set of test cases or someone with in-depth knowledge of the project to perform testing
+  - do we have established deployment slots (e.g. scheduled, manually approved)
+
+### Determining upgrade steps
+You should **always** update one minor version of Rails at a time. Each new minor version increment can introduce **important deprecation warnings** to warn you of changes in the next minor/major release. Skipping minor versions can lead to unexpected, hard to diagnose errors. It may seem like you're going slower one minor version at a time, but you'll be faster (and safer) in the long run. Minimizing the complexity and risk of code changes is always the better option.
```

Comment:
> You should **always** update one minor version of Rails at a time.

What about Ruby? I feel the same thing goes for Ruby & deprecation warnings

Reply:
I think we often cheat in practice with Ruby (e.g. 2.7.6 to 3.1 directly since we're late with upgrade cycles), but I agree we should at least strive to do better. The more up to date we are, the easier this becomes to implement in practice.